### PR TITLE
feat: Implement SignificantDigitsNumberRounder and IncrementNumberRounder RoundDouble Method

### PR DESCRIPTION
**Pull request by MohammadHadi2031**
GitHub Issue (If applicable): part of #6908

## PR Type

What kind of change does this PR introduce?

## What is the current behavior?

SignificantDigitsNumberRounder rounds double in a wrong way.
RoundDouble is not implement for IncrementNumberRounder.

## What is the new behavior?

Rounds correctly as UWP does for both SignificantDigitsNumberRounder and IncrementNumberRounder.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UWP/Globalization/NumberFormatting/INumberRounder.cs
Comment (Youssef1313) [inline src/Uno.UWP/Globalization/NumberFormatting/INumberRounder.cs:10]: nit: Keep formatting clean in non-generated file.

```suggestion
		int RoundInt32(int value);
		uint RoundUInt32(uint value);
		long RoundInt64(long value);
		ulong RoundUInt64(ulong value);
		float RoundSingle(float value);
		double RoundDouble(double value);
```

**Comment by MartinZikmund** [comment]
@MohammadHadi2031 Please sign the CLA 👍  
image

**Comment by MohammadHadi2031** [comment]
> @MohammadHadi2031 Please sign the CLA 👍  
> image
> 

I want to sign contribute licence agreement. But google shows 403(forbidden) error. It says your client does not have permission to get URL from this server. I used VPN but again same error has been shown. How can I solve this?

**Comment by Youssef1313** [comment]
> > @MohammadHadi2031 Please sign the CLA 👍
> > image
> 
> I want to sign contribute licence agreement. But google shows 403(forbidden) error. It says your client does not have permission to get URL from this server. I used VPN but again same error has been shown. How can I solve this?

Doesn't it open like this?

image
 
with a "Sign in with GitHub to agree" button at the bottom?

**Comment by MohammadHadi2031** [comment]
> > > @MohammadHadi2031 Please sign the CLA 👍
> > > image
> > 
> > 
> > I want to sign contribute licence agreement. But google shows 403(forbidden) error. It says your client does not have permission to get URL from this server. I used VPN but again same error has been shown. How can I solve this?
> 
> Doesn't it open like this?
> 
> image
> 
> with a "Sign in with GitHub to agree" button at the bottom?

No just show it.

image

**Comment by MohammadHadi2031** [comment]
> > > > @MohammadHadi2031 Please sign the CLA 👍
> > > > image
> > > 
> > > 
> > > I want to sign contribute licence agreement. But google shows 403(forbidden) error. It says your client does not have permission to get URL from this server. I used VPN but again same error has been shown. How can I solve this?
> > 
> > 
> > Doesn't it open like this?
> > image
> > with a "Sign in with GitHub to agree" button at the bottom?
> 
> No just show it.
> 
> image

I solved it.

**Comment by MohammadHadi2031** [comment]
@Youssef1313, Can I ask a favour? I don't know how to make these test pass. I didn't change anything in Uno.UI.RuntimeTests but Android/IOS runtime tests failed. Also they successfully passed at first CI workflows.

**Comment by MartinZikmund** [comment]
@MohammadHadi2031 It is likely that these failures are not actually caused by your changes, I will rerun and we shall see

**Comment by MohammadHadi2031** [comment]
I found a bug and I'm working on it. Please do not merge. Work in progress

**Comment by MohammadHadi2031** [comment]
> I found a bug and I'm working on it. Please do not merge. Work in progress

Fixed It.

File: src/Uno.UI.Tests/Windows_Globalization/Given_Rounder.cs
Comment (jeromelaban) [inline src/Uno.UI.Tests/Windows_Globalization/Given_Rounder.cs:7]: Nice PR !

Those tests are a good candidate to be running as runtime tests as well. You can reference globalization the two test files in here:

https://github.com/unoplatform/uno/blob/master/src/Uno.UI.RuntimeTests/Uno.UI.RuntimeTests.Wasm.csproj#L35-L48

and the other projects as well. We probably should extract those references into an additional `.props` file so we don't need to modify all `Uno.UI.RuntimeTests.*.csproj` files.

The interest here is to be able to run those same tests on windows to validate that they are behaving the same as our implementation.

Comment (MohammadHadi2031) [reply src/Uno.UI.Tests/Windows_Globalization/Given_Rounder.cs:7]: Thanks for your suggestion. I added references. Did I do it correctly?

**Comment by MohammadHadi2031** [comment]
@jeromelaban. I implemented RoundDouble for IncrementNumberRounder too. Please check it out  :)

**Comment by MohammadHadi2031** [comment]
I rebased my branch on master and now test fails.

2021-09-08T11:51:31.8947967Z Failed    Uno.UI.Tests.TextBoxTests.Given_TextBox.Calling_Select_With_In_Range_Values    Calling_Select_With_In_Range_Values
2021-09-08T11:51:31.8948925Z Failed    Uno.UI.Tests.TextBoxTests.Given_TextBox.Calling_Select_With_Out_Of_Range_Length    Calling_Select_With_Out_Of_Range_Length
2021-09-08T11:51:31.8949916Z Failed    Uno.UI.Tests.TextBoxTests.Given_TextBox.Calling_Select_With_Out_Of_Range_Start    Calling_Select_With_Out_Of_Range_Start

These aren't caused by my changes

**Comment by jeromelaban** [comment]
> These aren't caused by my changes

This is indeed not caused by your changes. We'll requeue a build.

**Comment by jeromelaban** [comment]
Thanks! This is all looking good! One additional validation, do your new tests pass when running the `Unit Tests` sample in the UWP sample app ?

**Comment by MohammadHadi2031** [comment]
> Thanks! This is all looking good! One additional validation, do your new tests pass when running the `Unit Tests` sample in the UWP sample app ?

Yeh. I actually have a visual studio solution file that has Net4.6.1 and UWP projects that is not part of my fork. The same test class is used for both. First I check the behavior in UWP then I go and write code for classes

**Comment by MohammadHadi2031** [comment]
If we run Given_SignificantDigitsNumberRounder and Given_IncrementNumberRounder with UWP as target platform we will be sure. Of course it was your suggestion but I don't know how.

**Comment by jeromelaban** [comment]
For that you'll need to build the sample app inside the uno solution using the  `Uno.UI-Windows-only.slnf` with the `uap10.0.17763` override.

**Comment by MohammadHadi2031** [comment]
> For that you'll need to build the sample app inside the uno solution using the  `Uno.UI-Windows-only.slnf` with the `uap10.0.17763` override.

Understood. Do we have a test in CI that runs with the `uap10.0.17763` override? It seems that we have UWP samples App but it doesn't run tests or am I wrong?

**Comment by MartinZikmund** [comment]
@MohammadHadi2031 Yes that is right, we don't run them as part of the CI. But you can validate at least locally, if the tests pass (mainly to verify the implementation's behavior matches)

File: src/Uno.UI.Tests/Windows_Globalization/Given_Rounder.cs
Comment (MartinZikmund) [inline src/Uno.UI.Tests/Windows_Globalization/Given_Rounder.cs:30]: The indentation is wrong for the remainder of the file, try format document in VS to adjust it 👍

File: src/Uno.UI.Tests/Windows_Globalization/Given_SignificantDigitsNumberRounder.cs
Comment (MartinZikmund) [inline src/Uno.UI.Tests/Windows_Globalization/Given_SignificantDigitsNumberRounder.cs:10]: Also here please, try formatting document

File: src/Uno.UWP/Globalization/NumberFormatting/IncrementNumberRounder.cs
Comment (MartinZikmund) [inline src/Uno.UWP/Globalization/NumberFormatting/IncrementNumberRounder.cs:30]: Please add curly braces here

File: src/Uno.UWP/Globalization/NumberFormatting/SignificantDigitsNumberRounder.cs
Comment (MartinZikmund) [inline src/Uno.UWP/Globalization/NumberFormatting/SignificantDigitsNumberRounder.cs:17]: The formatting seems to be off here

File: src/Uno.UWP/Globalization/NumberFormatting/Rounder.cs
Comment (MartinZikmund) [inline src/Uno.UWP/Globalization/NumberFormatting/Rounder.cs:5]: Please make this class `internal` as it is not part of the public UWP/WinUI contract. You can also move it to `Uno.Globalization.NumberFormatting` namespace

**Review by MartinZikmund — APPROVED** [review]

File: src/Uno.UWP/Globalization/NumberFormatting/Rounder.cs
Comment (MohammadHadi2031) [reply src/Uno.UWP/Globalization/NumberFormatting/Rounder.cs:5]: @MartinZikmund Is the file location correct? Where should I put the file?

**Comment by MohammadHadi2031** [comment]
> @MohammadHadi2031 Yes that is right, we don't run them as part of the CI. But you can validate at least locally, if the tests pass (mainly to verify the implementation's behavior matches)

I validate locally. Because the same test file is used for both UWP and Net4.6.1. And tests passed for both.

File: src/Uno.UI.Tests/Windows_Globalization/Given_Rounder.cs
Comment (MohammadHadi2031) [reply src/Uno.UI.Tests/Windows_Globalization/Given_Rounder.cs:31]: The problem was because of using spaces for indentation not Tabs.

**Comment by MohammadHadi2031** [comment]
Is there anything else that I should do for merging?

**Comment by MartinZikmund** [comment]
@MohammadHadi2031 No, needs one more approval, probably @jeromelaban will recheck, then will add `ready-to-merge` and will merge automatically 🙂

**Comment by jeromelaban** [comment]
I've changed a bit the runtime tests assembly so the new tests can run on UWP. I validated that they actually run properly.

image

**Review by jeromelaban — APPROVED** [review]

**Comment by MohammadHadi2031** [comment]
@jeromelaban Would you mind rerunning failed test? 🙂

**Pull request merged**
